sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -1990,6 +1990,90 @@ void ProgramManager::addImages(sycl_device_binaries DeviceBinary) {
   }
 }
 
+void ProgramManager::removeImages(sycl_device_binaries DeviceBinary) {
+  for (int I = 0; I < DeviceBinary->NumDeviceBinaries; I++) {
+    sycl_device_binary RawImg = &(DeviceBinary->DeviceBinaries[I]);
+    const sycl_offload_entry EntriesB = RawImg->EntriesBegin;
+    const sycl_offload_entry EntriesE = RawImg->EntriesEnd;
+    // Treat the image as empty one
+    if (EntriesB == EntriesE)
+      continue;
+
+    // Retrieve raw image by looking up the first offload entry
+    kernel_id FirstKernelID = getSYCLKernelID(RawImg->EntriesBegin->name);
+    auto RawImages = getRawDeviceImages({FirstKernelID});
+    assert(RawImages.size() == 1);
+
+    RTDeviceBinaryImage *Img = *RawImages.begin();
+
+    // Drop the kernel argument mask map
+    // TODO: Why is this not protected by a mutex?
```

Comment:
Do we still need to address this TODO as part of this PR?

Reply:
What the TODO is asking is why `m_EliminatedKernelArgMasks` not protected by a mutex in the existing code, in particular in `addImages` which my method here tries to mirror. So this is potentially a separate issue, and I was hoping reviewers could explain the rationale here.

Reply:
I don't see why it shouldn't be. seems like a bug to me.

Reply:
I filed https://github.com/intel/llvm/issues/16836, and dropped the TODO from the code for now.